fem/ceed/algebraic.cpp

```diff
@@ -0,0 +1,989 @@
+                                                         const mfem::Operator *P_)
+   : ConstrainedMFEMCeedOperator(oper, Array<int>(), P_)
+{ }
+
+ConstrainedMFEMCeedOperator::~ConstrainedMFEMCeedOperator()
+{
+   delete constrained_op;
+   delete unconstrained_op;
+}
+
+void ConstrainedMFEMCeedOperator::Mult(const Vector& x, Vector& y) const
+{
+   constrained_op->Mult(x, y);
+}
+
+CeedOperator ConstrainedMFEMCeedOperator::GetCeedOperator() const
+{
+   return unconstrained_op->GetCeedOperator();
+}
+
+const Array<int> &ConstrainedMFEMCeedOperator::GetEssentialTrueDofs() const
+{
+   return ess_tdofs;
+}
+
+const mfem::Operator *ConstrainedMFEMCeedOperator::GetProlongation() const
+{
+   return P;
+}
+
+/// assumes a square operator (you could do rectangular, you'd have
+/// to find separate active input and output fields/restrictions)
+int CeedOperatorGetSize(CeedOperator oper, CeedInt * size)
+{
+   int ierr;
+   CeedElemRestriction er;
+   ierr = CeedOperatorGetActiveElemRestriction(oper, &er); CeedChk(ierr);
+   ierr = CeedElemRestrictionGetLVectorSize(er, size); CeedChk(ierr);
+   return 0;
+}
+
+Solver *BuildSmootherFromCeed(ConstrainedMFEMCeedOperator &op, bool chebyshev)
+{
+   int ierr;
+   CeedOperator ceed_op = op.GetCeedOperator();
+   const Array<int> &ess_tdofs = op.GetEssentialTrueDofs();
+   const mfem::Operator *P = op.GetProlongation();
+   // Assemble the a local diagonal, in the sense of L-vector
+   CeedVector diagceed;
+   CeedInt length;
+   ierr = CeedOperatorGetSize(ceed_op, &length); PCeedChk(ierr);
+   ierr = CeedVectorCreate(internal::ceed, length, &diagceed); PCeedChk(ierr);
+   CeedMemType mem;
+   ierr = CeedGetPreferredMemType(internal::ceed, &mem); PCeedChk(ierr);
+   if (!Device::Allows(Backend::CUDA) || mem != CEED_MEM_DEVICE)
+   {
+      mem = CEED_MEM_HOST;
+   }
+   Vector local_diag(length);
+   CeedScalar *ptr = (mem == CEED_MEM_HOST) ? local_diag.HostWrite() :
+                     local_diag.Write(true);
+   ierr = CeedVectorSetArray(diagceed, mem, CEED_USE_POINTER, ptr);
+   PCeedChk(ierr);
+   ierr = CeedOperatorLinearAssembleDiagonal(ceed_op, diagceed,
+                                             CEED_REQUEST_IMMEDIATE);
+   PCeedChk(ierr);
+   ierr = CeedVectorTakeArray(diagceed, mem, NULL); PCeedChk(ierr);
+
+   Vector t_diag;
+   if (P)
+   {
+      t_diag.SetSize(P->Width());
+      P->MultTranspose(local_diag, t_diag);
+   }
+   else
+   {
+      t_diag.NewMemoryAndSize(local_diag.GetMemory(), length, false);
+   }
+   Solver *out = NULL;
+   if (chebyshev)
+   {
+      const int cheb_order = 3;
+      out = new OperatorChebyshevSmoother(&op, t_diag, ess_tdofs, cheb_order);
+   }
+   else
+   {
+      const double jacobi_scale = 0.65;
+      out = new OperatorJacobiSmoother(t_diag, ess_tdofs, jacobi_scale);
+   }
+   ierr = CeedVectorDestroy(&diagceed); PCeedChk(ierr);
+   return out;
+}
+
+#ifdef MFEM_USE_MPI
+
+class CeedAMG : public Solver
```

Comment:
I would rename to `AMG` or `AlgebraicMultigridSolver`.

Reply:
I renamed it `AssembledAMG` which I think describes its purpose.